### ISSUE TITLE
Galileo ID range

### ISSUE THREAD

**Issue by barbeau** [mvglasow/satstat#110]
I've confirmed Galileo ID range on at least one Android M or older device to be 301-330 (BQ Aquaris X5+, with Qualcomm chipset).

See http://gis.stackexchange.com/a/211229/12198.

I've added it to my app here, as well as Android N explicit support for Galileo and other GNSS:
https://play.google.com/store/apps/details?id=com.android.gpstest

**Comment by mvglasow** [comment]
Thanks! I'll add this to the next release.

**Comment by utack** [comment]
Can I somehow check if my device (theoretically) supports Galileo?

**Comment by barbeau** [comment]
Other than running an app that supports Galileo and see if it shows up, not that I know of.

**Comment by mvglasow** [comment]
I'm not aware of any way to query that via the API. In SatStat, I build the SNR graph by looking at satellite status reports and enable ranges based on the NMEA IDs I get. You can see that when you first launch SatStat – the SNR graph will show only GPS, other supported systems will show a few seconds later. On my device, the SBAS range occasionally shows when I pick up one of the sats in that range, other times it'll stay hidden because the device hasn't picked up any sats in that range, so it wouldn't know they're supported.

**Comment by barbeau** [comment]
I've noticed on Android N that SBAS are much more likely to show up.  I have an LG G5 that originally had 6.0, and no SBAS ever showed up in view.  However, with Android 7.0 (N update) I now see SBAS as well.  I think the OEMs are more comfortable exposing non-GPS sats via the Android N API that explicitly supports them, rather than cramming other GNSS constellations into the `getPrn()` method that was originally designed for GPS.

**Comment by utack** [comment]
Anyone know when this Galileo launch is exactly, so I can try if my device can receive them?

**Comment by mvglasow** [comment]
@barbeau Thanks for the research! I’ve added support for Galileo (pseudo PRN-based) in 5f10dd41. Since I have no Galileo-capable device, I can only do regression tests. If anyone feels like building this and testing, results would be appreciated. Otherwise this will go into the next F-Droid release as-is.

**Comment by ArchangeGabriel** [comment]
I don’t feel like building, but if you provide me a build I can test it on a Galileo-capable device.

**Comment by breversa** [comment]
So, how about a 3.3 release ? :-)

**Comment by mvglasow** [comment]
Under construction, I’m rounding up everything I want to go in there, and then I’ll put it out there.

**Comment by breversa** [comment]
Has anyone successfully tested SatStat with a Galileo-capable device ?

**Comment by ArchangeGabriel** [comment]
My above comment still stands.

**Comment by breversa** [comment]
I meant : "... now that v3.3 is out ?" :)

**Comment by mvglasow** [comment]
@ArchangeGabriel the [official 3.3 build](https://f-droid.org/packages/com.vonglasow.michael.satstat/) theoretically has support for Galileo ranges. If you can test this, results would be appreciated.

**Comment by ArchangeGabriel** [comment]
Hum indeed. Then it does not work for me.

**Comment by ArchangeGabriel** [comment]
There is just no Galileo range displayed. But I’m currently trying other GPS apps, and they show empty range for Galileo as well, so it might be an issue on my side.

**Comment by mvglasow** [comment]
If you compare the sky plot with the satellite ranges, does the number of satellites match? Or are you seeing more satellites in the sky plot than in the SNR view?

The sky plot shows all satellites reported by the API without any filtering, whereas the SNR graph displays only known ranges. If you are seeing more satellites in the sky plot than in the SNR graph, your phone is reporting satellites in a range SatStat does not know about.

**Comment by ArchangeGabriel** [comment]
Same number in both. But as I said above, it seems Galileo doesn’t work currently. I’m sure it did at some point because https://play.google.com/store/apps/details?id=com.android.gpstest reported Galileo satellites, but it does not currently.

**Comment by ArchangeGabriel** [comment]
Rectification: moving some meters away, I got two Galileo satellites to show up in GPStest. But still nothing special in SatStat. However they are a lot of fluctuations, and GPStest reports more satellites than SatStat.

**Comment by ArchangeGabriel** [comment]
OK, I’ve just remembered about the split screen feature of Android and managed to get screenshots showing the difference. They are heavily redacted, but you can see that there is one more total satellite listed by GPStest.

![screen1](https://user-images.githubusercontent.com/490531/41103771-743f5788-6a59-11e8-85a8-cf65b5213dbc.png)
![screen2](https://user-images.githubusercontent.com/490531/41103772-745d7f2e-6a59-11e8-9078-60083e0e7da3.png)

BTW the code is open for this app, so maybe you will find the solution there: https://github.com/barbeau/gpstest

**Comment by barbeau** [comment]
@ArchangeGabriel In GPSTest, if in Settings you uncheck "Use GNSS APIs", do the two apps then match?

Also, what's the make, model, and Android version of your device?

**Comment by ArchangeGabriel** [comment]
@barbeau Yes, unchecking this make both apps report the same thing, because the Galileo sat disappear right away when doing so.

OnePlus 5T, Android 8.1 (OmniROM, I might switch to LineageOS in the future if that matters).

**Comment by barbeau** [comment]
Ok, that makes sense.  

From a quick look [at the code](https://github.com/mvglasow/satstat/blob/4e0dcef1ee7aa986d90d71aec54f68ec7f972dcf/src/com/vonglasow/michael/satstat/ui/MainActivity.java) SatStat is using the [`GpsStatus.Listener`](https://developer.android.com/reference/android/location/GpsStatus) for satellite signal updates.  This format dates back to the early Android days and was GPS-specific.  Some OEMs chose to expose signals from non-GPS satellites via this API but other constellations were never officially supported (hence the manual PRN-to-constellation mapping that needed to be done to show non-GPS satellites).  

Android N introduced a new [`GnssStatus.Callback`](https://developer.android.com/reference/android/location/GnssStatus) which explicitly supports all GNSS, not just GPS - it has [a specific constant for Galileo](https://developer.android.com/reference/android/location/GnssStatus.html#CONSTELLATION_GALILEO) as well as other GNSS.  As a result, the old `GpsStatus.Listener` has been deprecated.

In GPSTest, when "Use GNSS APIs" is checked it's using the new `GnssStatus.Callback`, which has explicit support for Galileo, and therefore OEMs are providing Galileo information as required by Android N and higher.  When "Use GNSS APIs" is unchecked, it's using the legacy `GpsStatus.Listener`, and OEMs may or may not show Galileo info via this deprecated API - non-GPS constellations aren't officially supported.

So, the solution for Galileo to show up in SatStat on your device would be for SatStat to add support for `GnssStatus.Callback` on Android N and higher.

**Comment by ArchangeGabriel** [comment]
@barbeau Thank you very much for this detailed analysis and explanation. It definitively makes sense, and I guess @mvglasow has some work to do now. ;)

**Comment by barbeau** [comment]
Or, @ArchangeGabriel, *you* have some work to do now ;). Most of us that maintain these types of OSS projects do so in our free time, with the community pitching in to accelerate development as they are interested. It would be a great learn Android GNSS project. Just sayin' :).

**Comment by ArchangeGabriel** [comment]
Well, I don’t know a keyword of Java, and have no intention to learn any whatsoever. ;) And I already lack time to involve myself as much as I would like in the numerous FLOSS project I’m part of, so believe me, I know how it (does not…) works. :(
